File: app/schemas/paper.py
from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel

from app.schemas.question import QuestionResponse


class GeneratePaperRequest(BaseModel):
    prompt: str


class PaperQuestionItem(BaseModel):
    id: uuid.UUID                   # PaperQuestion row ID
    paper_id: uuid.UUID
    question_id: uuid.UUID
    position: int
    paper_section: str | None = None
    display_number: int | None = None
    locked: bool
    question: QuestionResponse

    model_config = {"from_attributes": True}


class PaperResponse(BaseModel):
    id: uuid.UUID
    prompt: str
    blueprint: dict
    status: str
    notes: str | None = None
    questions: list[PaperQuestionItem] = []
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class PaperListItem(BaseModel):
    id: uuid.UUID
    prompt: str
    status: str
    created_at: datetime

    model_config = {"from_attributes": True}


class SwapQuestionRequest(BaseModel):
    new_question_id: uuid.UUID


class AddQuestionRequest(BaseModel):
    question_id: uuid.UUID
    position: int | None = None     # If None, appends to end
    paper_section: str | None = None


class ApproveRequest(BaseModel):
    notes: str | None = None


class AlternativesResponse(BaseModel):
    alternatives: list[QuestionResponse]

File: admin-panel/src/types/paper.ts
// Types for the paper generation workflow

import type { QuestionDetail } from './question'

// ── Syllabus Query (natural language → draft paper) ────────────────────────

export interface QuestionPreview {
  id: string
  question_number: number | null
  chapter: string | null
  difficulty: string | null
  question_type: string | null
  stem_preview: string
}

export interface ResolvedBlueprint {
  exam_name: string
  subject: string
  resolved_chapters: string[]
  excluded_chapters: string[]
  chapter_filter_mode: string
  question_count: number
  difficulty: string | null
  difficulty_distribution: Record<string, number> | null
  question_type: string | null
  section_type: string | null
  has_formula: boolean | null
  has_diagram: boolean | null
  concepts: string[]
  concept_match_mode: string
  chapter_range_description: string
  resolver_warnings: string[]
}

export interface CandidateSet {
  total_available: number
  by_chapter: Record<string, number>
  by_difficulty: Record<string, number>
  by_section_type: Record<string, number>
  by_question_type: Record<string, number>
  sample_questions: QuestionPreview[]
  can_generate: boolean
  shortage: number
  resolved_blueprint: ResolvedBlueprint
}

export interface SyllabusQueryResponse {
  paper_id: string | null
  candidates: CandidateSet
  paper: PaperSummary | null
}

// ── Paper ──────────────────────────────────────────────────────────────────

export interface PaperQuestionItem {
  id: string           // PaperQuestion row ID
  paper_id: string
  question_id: string
  position: number
  paper_section: string | null
  display_number: number | null
  locked: boolean
  question: QuestionDetail
}

export interface Paper {
  id: string
  prompt: string
  blueprint: Record<string, unknown>
  status: string        // "draft" | "approved"
  notes: string | null
  questions: PaperQuestionItem[]
  created_at: string
  updated_at: string
}

export interface PaperSummary {
  id: string
  prompt: string
  status: string
  created_at: string
}

export interface AlternativesResponse {
  alternatives: QuestionDetail[]
}
